server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1039,11 +1039,19 @@ private VersionValue getVersionFromMap(BytesRef id) {
                 // but we only need to do this once since the last operation per ID is to add to the version
                 // map so once we pass this point we can safely lookup from the version map.
                 if (versionMap.isUnsafe()) {
-                    lastUnsafeSegmentGenerationForGets.set(lastCommittedSegmentInfos.getGeneration() + 1);
```

Comment:
I think we need both and wonder if this relates to the issues we see now? Otherwise, we could grab the last unsafe commit generation prior to setting it below, causing issues.

Reply:
What issue you are referring to? Is it the test failure that has NoSuchFileException as its cause? If so, I can be certain it is not related because we observed that failure before this got merged.

Reply:
I think we should move this into either before/after-refresh or into `flush`. That would put it under appropriate locks, ensuring that other refresh or flush'es do not happen.

Reply:
Is the concern that we could be setting `lastUnsafeSegmentGenerationForGets` too high sometimes?

Reply:
I discussed this further with Pooya and I am now convinced that it is safe enough, since:

1. If versionMap.isUnsafe() == false, we know that the search shard is up to date with any unsafe information (since the archive will otherwise force it to true).
2. Otherwise, we pass the code here and it forces a flush and for RTG to wait for the flushed generation.

Reply:
Thanks both for the discussion. Sorry that I was outside and could not comment earlier. I'd like to add that I am reasonbly confident that the change is safe as is. It is not the most efficient since the search shard could wait for extra generations. But even this extra wait is safe in that the extra generations are guaranteed to happen.

I also think having this discussion is a sign that we do not have tests that explicitly and precisely excercise all scenarios. We do have tests that with enough randomness will cover all scenarios. But it would be great if we can have a way to test different scenarios with precision, e.g. pausing right after `commitIndexWriter` and start new indexing etc. I think we might be able to write such tests by subclassing/delegating the `IndexEngine`. I'll see what I can do.